pkg/cmd/p2p/promote/p2p_promote.go

```diff
@@ -0,0 +1,161 @@
+package promote
+
+import (
+	"fmt"
+	"github.com/coreeng/corectl/pkg/command"
+	"github.com/spf13/cobra"
+	"io"
+	"os"
+	"strings"
+)
+
+type promoteOpts struct {
+	ImageWithTag       string
+	SourceRegistry     string
+	SourceStage        string
+	SourceAuthOverride string
+	DestRegistry       string
+	DestStage          string
+	DestAuthOverride   string
+	Exec               command.Commander
+	Out                io.Writer
+}
+
+type imageOpts struct {
+	ImageNameWithTag string
+	Registry         string
+	RepoPath         string
+	AuthOverride     string
+}
+
+func NewP2PPromoteCmd() (*cobra.Command, error) {
+	var opts = promoteOpts{
+		Exec: command.NewCommand(),
+	}
+	var promoteCommand = &cobra.Command{
+		Use:   "promote <image_with_tag>",
+		Short: "Promotes image from source to destination registry. Only GCP is supported for now",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.ImageWithTag = args[0]
+			opts.Out = cmd.OutOrStdout()
+			return run(&opts)
+		},
+	}
+
+	requiredFlagMap := map[string]*string{
+		"source-registry": &opts.SourceRegistry,
+		"source-stage":    &opts.SourceStage,
+		"dest-registry":   &opts.DestRegistry,
+		"dest-stage":      &opts.DestStage,
+	}
+
+	for name, field := range requiredFlagMap {
+		err := addFlag(promoteCommand, field, name, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	optionalFlagMap := map[string]*string{
+		"source-auth-override": &opts.SourceAuthOverride,
+		"dest-auth-override":   &opts.DestAuthOverride,
+	}
+	for name, field := range optionalFlagMap {
+		err := addFlag(promoteCommand, field, name, false)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return promoteCommand, nil
+}
+
+func addFlag(promoteCommand *cobra.Command, field *string, name string, required bool) error {
+	envVariableName := strings.ToUpper(strings.ReplaceAll(name, "-", "_"))
+
+	description := fmt.Sprintf("optional, defaults to environment variable: %s", envVariableName)
+	if required {
+		description = fmt.Sprintf("required, defaults to environment variable: %s", envVariableName)
+	}
+
+	promoteCommand.Flags().StringVar(
+		field,
+		name,
+		"",
+		description,
+	)
+
+	envVariableValue := os.Getenv(envVariableName)
+
+	if envVariableValue != "" {
+		*field = envVariableValue
+	}
+
+	if envVariableValue == "" && required {
+		if err := promoteCommand.MarkFlagRequired(name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func run(opts *promoteOpts) error {
+	if err := validate(opts); err != nil {
+		return err
+	}
+
+	sourceImage := &imageOpts{
+		ImageNameWithTag: opts.ImageWithTag,
+		Registry:         opts.SourceRegistry,
+		RepoPath:         opts.SourceStage,
+		AuthOverride:     opts.SourceAuthOverride,
+	}
+
+	destinationImage := &imageOpts{
```

Comment:
disagree, it would break readability of the precedure down below. Matter of taste, I like it this way ;) 
![CleanShot 2024-07-29 at 13 08 37@2x](https://github.com/user-attachments/assets/66ec81c3-2a5d-49e3-a566-baa8e02fe1eb)